src/Sia.Gateway/Hubs/EventsHub.cs

```diff
@@ -14,15 +18,90 @@ namespace Sia.Gateway.Hubs
     public class EventsHub : Hub
     {
         public const string HubPath = "events/live";
-        public EventsHub() : base()
+        private readonly ConcurrentDictionary<string, IFilterByMatch<Event>> _filterLookup;
+        private readonly ILogger<EventsHub> _logger;
+
+        public EventsHub(
+            ConcurrentDictionary<string, IFilterByMatch<Event>> filterLookup,
+            ILoggerFactory loggerFactory
+        ) : base()
         {
+            _filterLookup = filterLookup;
+            _logger = loggerFactory.CreateLogger<EventsHub>();
+        }
 
+        public override async Task OnConnectedAsync()
+        {
+            await base.OnConnectedAsync();
+            _logger.LogInformation(
+                "User {0} connected to EventsHub on connectionId {1}",
+                new object[] { Context.User.Identity.Name, Context.ConnectionId }
+            );
+            await ClearFilter();
+        }
+
+        public override async Task OnDisconnectedAsync(Exception exception)
+        {
+            await ClearFilter();
+            
+            if(exception == null)
+            {
+                _logger.LogInformation(
+                    "User {0} disconnected from EventsHub on connectionId {1} without exception",
+                    new object[] { Context.User.Identity.Name, Context.ConnectionId }
+                );
+            }
+            else
+            {
+                _logger.LogError(
+                    exception,
+                    "User {0} disconnected from EventsHub on connectionId {1} with exception; see exception for details.",
+                    new object[] { Context.User.Identity.Name, Context.ConnectionId }
+                );
+            }
+
+            await base.OnDisconnectedAsync(exception);
         }
 
         public Task Send(Event ev)
+            => Clients.AllExcept(
+                    _filterLookup
+                        .Where((kvp) => !kvp.Value.IsMatchFor(ev))
+                        .Select((kvp) => kvp.Key)
+                        .Append(Context.ConnectionId) // Prevent loops (if those are even possible)
+                        .ToList()
+                ).InvokeAsync("Send", Json(ev));
+
+        public Task UpdateFilter(EventFilters filters)
```

Comment:
I think this can be rephrased to
```
public async Task UpdateFilter(EventFilters filters)
{
    ...
    // No need to return Task.CompletedTask;
}
```

Reply:
This one also doesn't need to return a task, it works as a void. Changed it to void.